File: src/cacheManager.ts
import * as fs from 'fs';
import * as os from 'os';
import * as path from 'path';
import {
  IAppImage,
  IFlathub,
  ICacheManager,
  IConfig,
  IUpdateCacheObject,
  ISnap,
} from './dataStructure';
import { successfullMessage } from './helpers';


/**
 * Cache helper class for managing the cache
 * @class CacheManager
 */
export class CacheManager implements ICacheManager {

  public isCacheEnabled: boolean;
  public hasCachedSources: boolean;
  private readonly cacheLocation: string;
  private readonly flathubCachePath: string;
  private readonly snapCachePath: string;
  private readonly appimageCachePath: string;
  private readonly updatedAtFilePath: string;
  private readonly updateCacheInterval: number;
  private readonly configPath: string;
  private readonly config: IConfig;



  constructor() {
    this.cacheLocation = path.resolve(os.homedir(), 'chob');
    this.flathubCachePath = path.resolve(this.cacheLocation, '.flathub.json');
    this.snapCachePath = path.resolve(this.cacheLocation, '.snap.json');
    this.appimageCachePath = path.resolve(this.cacheLocation, '.appimage.json');
    this.updatedAtFilePath = path.resolve(this.cacheLocation, '.updatedAt');
    this.configPath = path.resolve(this.cacheLocation, 'config.json');


    if (!fs.existsSync(this.cacheLocation)) {
      fs.mkdirSync(this.cacheLocation);
    }

    if (!this.checkCacheFiles()) {
      this.createCacheFiles();
    }

    this.config = this.parseConfig();
    this.isCacheEnabled = this.config.cacheEnabled;
    this.updateCacheInterval = this.config.chacheUpdateInterval || 1;


    this.hasCachedSources = this.checkHasCachedSources();


  }

  public updateInterval(interval: number): boolean {
    this.config.chacheUpdateInterval = interval;

    try {
      fs.writeFileSync(this.configPath, JSON.stringify(this.config));

      return true;
    } catch (err) {
      return false;
    }
  }

  public updateCacheStatment(statment: boolean): boolean {
    this.config.cacheEnabled = statment;
    try {
      fs.writeFileSync(this.configPath, JSON.stringify(this.config));
      return true;
    } catch (err) {
      return false;
    }
  }

  public getSourcesFromCache(): IUpdateCacheObject {
    const flathubData: IFlathub[] = JSON.parse(fs.readFileSync(this.flathubCachePath).toString());
    const snapData: ISnap = JSON.parse(fs.readFileSync(this.snapCachePath).toString());
    const appimageData: IAppImage = JSON.parse(fs.readFileSync(this.appimageCachePath).toString());

    return {
      flathubData,
      snapData,
      appimageData,
    };
  }


  public updateCache(data: IUpdateCacheObject): boolean | Error {
    try {
      const { flathubData, snapData, appimageData } = data;

      fs.writeFileSync(this.flathubCachePath, JSON.stringify(flathubData));
      fs.writeFileSync(this.snapCachePath, JSON.stringify(snapData));
      fs.writeFileSync(this.appimageCachePath, JSON.stringify(appimageData));
      fs.writeFileSync(this.updatedAtFilePath, Date.now().toString());

      successfullMessage('⚡ Created cache, your results will now pull from cached sources! ⚡');

      return true;

    } catch (err) {
      throw new Error(err);
    }
  }

  public shouldUpdateCache() {
    const date = Number(fs.readFileSync(this.updatedAtFilePath).toString());
    const tomorrowDate = new Date(date).setDate(new Date(date).getDate() + this.updateCacheInterval);

    if (!date) {
      return true;
    }

    if (tomorrowDate < Date.now()) {
      return true;
    }

    return false;


  }

  private parseConfig(): IConfig {
    const configFile = fs.readFileSync(this.configPath).toString();

    return JSON.parse(configFile);
  }

  private checkHasCachedSources(): boolean {
    const flathubData: IFlathub[] = JSON.parse(fs.readFileSync(this.flathubCachePath).toString());

    if (Object.keys(flathubData).length > 0) {
      return true;
    }

    return false;

  }

  private checkCacheFiles(): boolean {
    const flathubFile = fs.existsSync(this.flathubCachePath);
    const snapFile = fs.existsSync(this.snapCachePath);
    const appimageFile = fs.existsSync(this.appimageCachePath);
    const updateFile = fs.existsSync(this.updatedAtFilePath);
    const configPath = fs.existsSync(this.configPath);

    if (flathubFile && snapFile && appimageFile && updateFile && configPath) {
      return true;
    }

    return false;

  }

  private createCacheFiles(): boolean | Error {
    try {
      const data = JSON.stringify({});
      fs.writeFileSync(this.flathubCachePath, data);
      fs.writeFileSync(this.snapCachePath, data);
      fs.writeFileSync(this.appimageCachePath, data);
      fs.writeFileSync(this.updatedAtFilePath, '');
      fs.writeFileSync(this.configPath, data);


      return true;
    } catch (err) {
      throw new Error('Could not create cache files!');
    }
  }


}

File: src/cache_manager.py
from os import path, makedirs, stat
from pathlib import Path
from json import dumps, loads
from typing import Dict, List, TypeVar, TypedDict, Union, cast, no_type_check_decorator
from type.snap import Snap
from type.flathub import FlatHub
from type.appimage import AppImage
from dataclasses import dataclass


class Config(TypedDict):
    cache_enabled: bool
    cache_update_interval: int


@dataclass
class UpdateCache(TypedDict):
    flathub_data: List[FlatHub]
    snap_data: List[Snap]
    appimage_data: List[AppImage]


class CacheManager:
    is_cache_enabled: bool
    has_cached_sources: bool
    __cache_location: str
    __flathub_cache_path: str
    __snap_cache_path: str
    __appimage_cache_path: str
    __updatedAt_cache_path: str
    __config_path: str
    __config: Config
    __update_cache_interval: int

    def __init__(self) -> None:
        self.__cache_location = path.join(Path.home(), 'chob')
        self.__flathub_cache_path = path.join(
            self.__cache_location, '.flathub.json'
        )
        self.__snap_cache_path = path.join(self.__cache_location, '.snap.json'
                                           )
        self.__appimage_cache_path = path.join(
            self.__cache_location, '.appimage.json'
        )
        self.__updatedAt_cache_path = path.join(
            self.__cache_location, '.updatedAt'
        )
        self.__config_path = path.join(self.__cache_location, 'config.json')

        if not path.exists(self.__cache_location):
            makedirs(self.__cache_location)

        if not self.__check_cache_files():
            self.__create_cache_files()

        self.__parse_config()
        self.__update_cache_interval = self.__config.get(
            'cache_update_interval') or 1
        self.is_cache_enabled = self.__config.get('cache_enabled') or True
        self.has_cached_sources = self.__check_has_cached_sources()

    def fetch_sources_from_cache(self) -> UpdateCache:
        flathub_data: List[FlatHub]
        snap_data: List[Snap]
        appimage_data: List[AppImage]

        with open(self.__flathub_cache_path, 'r') as f:
            flathub_data = cast(List[FlatHub], loads(f.read()))
            f.close()

        with open(self.__snap_cache_path, 'r') as f:
            snap_data = cast(List[Snap], loads(f.read()))
            f.close()

        with open(self.__appimage_cache_path, 'r') as f:
            appimage_data = cast(List[AppImage], loads(f.read()))
            f.close()

        return {
            "flathub_data": flathub_data,
            "snap_data": snap_data,
            "appimage_data": appimage_data
        }

    def update_cache_statment(self, statment: bool) -> bool:
        self.__config['cache_enabled'] = statment

        try:
            with open(self.__config_path, 'w') as f:
                f.write(dumps(self.__config))
                f.close()
                return True
        except Exception:
            return False

    def update_cache(self, data: UpdateCache) -> bool:
        try:
            with open(self.__flathub_cache_path, 'w') as f:
                f.write(dumps(data['flathub_data']))
                f.close()

            with open(self.__snap_cache_path, 'w') as f:
                f.write(dumps(data['snap_data']))
                f.close()

            with open(self.__appimage_cache_path, 'w') as f:
                f.write(dumps(data['appimage_data']))
                f.close()

            return True
        except Exception:
            return False

    def __check_has_cached_sources(self) -> bool:
        with open(self.__flathub_cache_path, 'r') as f:
            data = loads(f.read())

            if len(data) > 0:
                f.close()
                return True

            f.close()
            return False

    def __parse_config(self) -> Config:
        with open(self.__config_path, 'r') as f:
            try:
                self.__config = loads(f.read())
                f.close()

                return self.__config
            except Exception:
                raise Exception('Config file could be corrupted! Remove it.')

    def __check_cache_files(self) -> bool:
        flathub_file: bool = path.exists(self.__flathub_cache_path)
        snap_file: bool = path.exists(self.__snap_cache_path)
        appimage_file: bool = path.exists(self.__appimage_cache_path)
        updated_at_file: bool = path.exists(self.__updatedAt_cache_path)
        config_file: bool = path.exists(self.__config_path)

        if flathub_file and snap_file and appimage_file and updated_at_file and config_file:
            return True

        return False

    def __create_cache_files(self) -> bool:
        files = [
            self.__flathub_cache_path,
            self.__snap_cache_path,
            self.__appimage_cache_path,
            self.__updatedAt_cache_path,
            self.__config_path
        ]

        data = dumps({})
        try:
            for file in files:
                with open(file, 'w') as f:
                    f.write(data)
                    f.close()
            return True
        except Exception:
            raise FileExistsError('Could not create cache files!')
